### PR TITLE
jujushell: handle possible initial welcome message.

### PR DESCRIPTION
**Pull request by frankban**
Also add support for the new jujushell API, which allows a more solid check that a response is related to a specific operation and also allows arbitrary welcome messages to be sent from the server.

### CODE REVIEW

**Comment by frankban** [comment]
The server change is at https://github.com/juju/jujushell/pull/12
The two branches must be used together in order to QA.
- start the shell with `jujushell cmd/jujushell/config-jaas-insecure.yam`;
- start guiproxy with `guiproxy -env prod -flags terminal -config '"jujushellURL": "ws://dev.local:8047/ws/"'` where dev.local is the host in which jujushell is running;
- check that the welcome message is properly displayed.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8527/

File: jujugui/static/gui/src/app/components/terminal/terminal.js [hunk 1]
```diff
@@ -254,4 +273,10 @@ Terminal.propTypes = {
   })
 };
 
+// Define jujushell API operations and codes.
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/terminal/terminal.js]: these constants should probably be defined at the top of the file.

File: jujugui/static/gui/src/app/components/terminal/terminal.js [hunk 2]
```diff
@@ -135,6 +138,22 @@ class Terminal extends React.Component {
     this.attachResizeListener();
   }
 
+  /**
+    Write a welcome message to the given term instance.
+    @param {Object} term The xterm instance.
+    @param {String} message The optional response message from the server.
+  */
+  _welcome(term, message) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/terminal/terminal.js]: `_writeWelcomeMessage` ?

File: jujugui/static/gui/src/app/components/terminal/terminal.js [hunk 3]
```diff
@@ -112,9 +112,12 @@ class Terminal extends React.Component {
             }
           }
       }
-      if (resp.code === 'ok' && resp.message === 'session is ready') {
+      if (resp.code === CODE_OK && (
+        resp.message === 'session is ready' || // Legacy API.
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/terminal/terminal.js]: It's not even 'released' yet and we're already supporting a legacy API? Lets just delete this.

File: jujugui/static/gui/src/app/components/terminal/terminal.js (hunk 1 above)
Comment (frankban) [reply jujugui/static/gui/src/app/components/terminal/terminal.js]: I am not sure, they are not exported nor important, so giving them more precedence over important stuff seems arbitrary.

File: jujugui/static/gui/src/app/components/terminal/terminal.js (hunk 2 above)
Comment (frankban) [reply jujugui/static/gui/src/app/components/terminal/terminal.js]: AFAIK welcome can be used as a transitive verb, so I'd say I like _welcome more, but maybe I am just being more influenced by Go than Java...

File: jujugui/static/gui/src/app/components/terminal/terminal.js (hunk 3 above)
Comment (frankban) [reply jujugui/static/gui/src/app/components/terminal/terminal.js]: Well, we made a blog post describing how to set up your own shell, and we know people that did it, so, given how easy is to be backward compatible in this case, I'd say let's at least for a while. What do you think?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/juju-gui/8533/

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**